File: typescript/examples/info_candles.ts
#!/usr/bin/env npx ts-node
// @ts-nocheck
/**
 * Historical Candles Example
 *
 * Shows how to fetch historical candlestick (OHLCV) data.
 *
 * Note: candleSnapshot may not be available on all QuickNode endpoints.
 * Check the QuickNode docs for method availability.
 *
 * Usage:
 *     export QUICKNODE_ENDPOINT="https://your-endpoint.hype-mainnet.quiknode.pro/YOUR_TOKEN"
 *     npx ts-node info_candles.ts
 */

import { HyperliquidSDK } from '@quicknode/hyperliquid-sdk';

const ENDPOINT = process.env.QUICKNODE_ENDPOINT;

if (!ENDPOINT) {
  console.error("Set QUICKNODE_ENDPOINT environment variable");
  process.exit(1);
}

async function main() {
  // Single SDK instance — access everything through sdk.info, sdk.core, sdk.evm, etc.
  const sdk = new HyperliquidSDK(ENDPOINT);
  const info = sdk.info;

  console.log("=".repeat(50));
  console.log("Historical Candles");
  console.log("=".repeat(50));

  // Last 24 hours
  const now = Date.now();
  const dayAgo = now - (24 * 60 * 60 * 1000);

  // Fetch BTC 1-hour candles
  console.log("\n1. BTC 1-Hour Candles (last 24h):");
  try {
    const candles = await info.candles("BTC", "1h", dayAgo, now);
    console.log(`   Retrieved ${candles.length} candles`);
    if (candles.length > 0) {
      for (const c of candles.slice(-3)) {
        console.log(`   O:${c.o} H:${c.h} L:${c.l} C:${c.c}`);
      }
    }
  } catch (e) {
    console.log(`   Error: ${e}`);
    console.log("   Note: candleSnapshot may not be available on this endpoint");
  }

  // Predicted funding rates (supported on QuickNode)
  console.log("\n2. Predicted Funding Rates:");
  try {
    const fundings = await info.predictedFundings();
    console.log(`   ${fundings.length} assets with funding rates:`);
    // Structure: [[coin, [[source, {fundingRate, ...}], ...]], ...]
    for (const item of fundings.slice(0, 5)) {
      if (Array.isArray(item) && item.length >= 2) {
        const coin = item[0];
        const sources = item[1];
        if (sources && Array.isArray(sources) && sources.length > 0) {
          // Get HlPerp funding rate if available
          for (const src of sources) {
            if (Array.isArray(src) && src.length >= 2 && src[0] === "HlPerp") {
              const rate = parseFloat(src[1].fundingRate || "0") * 100;
              console.log(`   ${coin}: ${rate.toFixed(4)}%`);
              break;
            }
          }
        }
      }
    }
  } catch (e) {
    console.log(`   Error: ${e}`);
  }

  console.log("\n" + "=".repeat(50));
  console.log("Done!");
}

main().catch(console.error);

File: python/examples/info_candles.py
#!/usr/bin/env python3
"""
Historical Candles Example

Shows how to fetch historical candlestick (OHLCV) data.

Note: candle_snapshot may not be available on all QuickNode endpoints.
Check the QuickNode docs for method availability.

Requirements:
    pip install hyperliquid-sdk

Usage:
    export ENDPOINT="https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN"
    python info_candles.py
"""

import os
import sys
import time

from hyperliquid_sdk import HyperliquidSDK

ENDPOINT = os.environ.get("ENDPOINT") or os.environ.get("QUICKNODE_ENDPOINT")

if not ENDPOINT:
    print("Error: Set ENDPOINT environment variable")
    print("  export ENDPOINT='https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN'")
    sys.exit(1)


def main():
    print("=" * 50)
    print("Historical Candles")
    print("=" * 50)

    # Single SDK instance - access everything through sdk.info(), sdk.core(), sdk.evm()
    sdk = HyperliquidSDK(ENDPOINT)
    info = sdk.info()

    # Last 24 hours
    now = int(time.time() * 1000)
    day_ago = now - (24 * 60 * 60 * 1000)

    # Fetch BTC 1-hour candles
    print("\n1. BTC 1-Hour Candles (last 24h):")
    try:
        candles = info.candles("BTC", "1h", day_ago, now)
        print(f"   Retrieved {len(candles)} candles")
        if len(candles) > 0:
            for c in candles[-3:]:
                print(f"   O:{c.get('o')} H:{c.get('h')} L:{c.get('l')} C:{c.get('c')}")
    except Exception as e:
        print(f"   Error: {e}")
        print("   Note: candle_snapshot may not be available on this endpoint")

    # Predicted funding rates (supported on QuickNode)
    print("\n2. Predicted Funding Rates:")
    try:
        fundings = info.predicted_fundings()
        print(f"   {len(fundings)} assets with funding rates:")
        # Structure: [[coin, [[source, {fundingRate, ...}], ...]], ...]
        count = 0
        for item in fundings:
            if count >= 5:
                break
            if isinstance(item, list) and len(item) >= 2:
                coin = item[0]
                sources = item[1]
                if sources and isinstance(sources, list) and len(sources) > 0:
                    # Get HlPerp funding rate if available
                    for src in sources:
                        if isinstance(src, list) and len(src) >= 2 and src[0] == "HlPerp":
                            rate = float(src[1].get("fundingRate", "0")) * 100
                            print(f"   {coin}: {rate:.4f}%")
                            count += 1
                            break
    except Exception as e:
        print(f"   Error: {e}")

    print()
    print("=" * 50)
    print("Done!")


if __name__ == "__main__":
    main()
